File: Lecture09/074_fp_vs_oo_decomposition.rb
# Programming Languages
# Lecture 09: OOP vs. Functional Decomposition 

# Note: If Exp and Value are empty classes, we do not need them in a
# dynamically typed language, but they help show the structure and they
# can be useful places for code that applies to multiple subclasses.

class Exp
    # could put default implementations or helper methods here
  
end
  
class Value < Exp
    # this is overkill here, but is useful if you have multiple kinds of
    # /values/ in your language that can share methods that do not make sense 
    # for non-value expressions
end
  
class Int < Value
    attr_reader :i
    
    def initialize i
        @i = i
    end

    def eval
        self
    end
    
    def to_s
        @i.to_s
    end

    def has_zero?
        i == 0
    end
end
  
class Negate < Exp
    attr_reader :e
    
    def initialize e
        @e = e
    end

    def eval
        Int.new(-e.eval.i) # error if e.eval has no i method
    end

    def to_s
        "-(" + e.to_s + ")"
    end

    def has_zero?
        e.has_zero?
    end
end
  
class Add < Exp
    attr_reader :e1, :e2

    def initialize(e1, e2)
        @e1 = e1
        @e2 = e2
    end

    def eval 
        Int.new(e1.eval.i + e2.eval.i) # error if e1.eval or e2.eval has no i method
    end

    def to_s
        "(" + e1.to_s + " + " + e2.to_s + ")"
    end

    def has_zero?
        e1.has_zero? || e2.has_zero?
    end
end

# 5 + 10 + (-3)
exp = Add.new(Int.new(5), Add.new(Int.new(10), Negate.new(Int.new(3))))

puts "Does exp have zero? " + exp.has_zero?.to_s
puts exp.to_s + " = " + exp.eval.to_s

File: Lecture09/074_fp_vs_oo_decomposition.cpp
// Programming Languages
// Lecture 09: OOP vs Functional Decomposition

#include <iostream>
#include <memory>
#include <string>

class Value;

class Exp
{
public:
    virtual std::shared_ptr<Value> eval() noexcept = 0;

    virtual std::string to_string() const noexcept = 0;

    virtual bool has_zero() const noexcept = 0;
};

class Value: public Exp
{

};

class Int: public Value
{
public:
    Int(int _i) noexcept
        : i{_i}
    {
        // Empty
    }

    std::shared_ptr<Value> eval() noexcept override
    {
        return std::make_shared<Int>(i);
    }

    std::string to_string() const noexcept override
    {
        return std::to_string(i);
    }

    bool has_zero() const noexcept override
    {
        return i == 0;
    }

public:
    int i;
};

class Negate: public Exp
{
public:
    Negate(std::shared_ptr<Exp> _e) noexcept
        : e{_e}
    {
        // Empty
    }

    std::shared_ptr<Value> eval() noexcept override
    {
        return std::make_shared<Int>(-std::dynamic_pointer_cast<Int>(e->eval())->i);
    }

    std::string to_string() const noexcept override
    {
        return "-(" + e->to_string() + ")";
    }

    bool has_zero() const noexcept override
    {
        return e->has_zero();
    }

private:
    std::shared_ptr<Exp> e;
};

class Add: public Exp
{
public:
    Add(std::shared_ptr<Exp> _e1, std::shared_ptr<Exp> _e2) noexcept
        : e1{_e1}, e2{_e2}
    {
        // Empty
    }

    std::shared_ptr<Value> eval() noexcept override
    {
        return std::make_shared<Int>(std::dynamic_pointer_cast<Int>(e1->eval())->i + std::dynamic_pointer_cast<Int>(e2->eval())->i);
    }

    std::string to_string() const noexcept override
    {
        return "(" + e1->to_string() + " + " + e2->to_string() + ")";
    }

    bool has_zero() const noexcept override
    {
        return e1->has_zero() || e2->has_zero();
    }

private:
    std::shared_ptr<Exp> e1;
    std::shared_ptr<Exp> e2;
};

int main()
{
    // 5 + 10 + (-3)
    auto e = std::make_shared<Add>(
        std::make_shared<Int>(5),
        std::make_shared<Add>(
            std::make_shared<Int>(10),
            std::make_shared<Negate>(std::make_shared<Int>(3))
        )
    );

    std::cout << std::boolalpha;
    std::cout << "Does e have zero? " << e->has_zero() << "\n";
    std::cout << e->to_string() << " = " << e->eval()->to_string() << "\n";
    return EXIT_SUCCESS;
}